File: frontend/src/lib/schemas.ts
import { z } from 'zod';

export const KeywordSchema = z.object({
  keyword_id: z.string().uuid(),
  value: z.string().min(1).max(64),
  normalized_value: z.string(),
  status: z.enum(['active', 'inactive']),
  created_at: z.string(),
});

export const KeywordListResponseSchema = z.object({
  items: z.array(KeywordSchema),
  total: z.number(),
});

export const MatchSchema = z.object({
  id: z.string().uuid(),
  keyword_id: z.string().uuid(),
  keyword_value: z.string(),
  certificate_sha256: z.string(),
  matched_field: z.enum(['cn', 'san', 'both']),
  matched_value: z.string(),
  domain_name: z.string().nullable(),
  issuer_cn: z.string().nullable(),
  issuer_org: z.string().nullable(),
  subject_cn: z.string().nullable(),
  subject_org: z.string().nullable(),
  not_before: z.string().nullable(),
  not_after: z.string().nullable(),
  first_seen_at: z.string(),
  last_seen_at: z.string(),
  is_new: z.boolean(),
  ct_log_index: z.number(),
});

export const MatchListResponseSchema = z.object({
  items: z.array(MatchSchema),
  total: z.number(),
});

export const MonitorStatusSchema = z.object({
  state: z.enum(['idle', 'running', 'error']),
  last_run_at: z.string().nullable(),
  last_success_at: z.string().nullable(),
  last_error_code: z.string().nullable(),
  last_error_message: z.string().nullable(),
  metrics_last_run: z.object({
    processed_count: z.number(),
    match_count: z.number(),
    parse_error_count: z.number(),
    duration_ms: z.number(),
    ct_latency_ms: z.number(),
    db_latency_ms: z.number(),
  }).nullable(),
});

export const CreateKeywordSchema = z.object({
  value: z.string().min(1).max(64),
});

export type Keyword = z.infer<typeof KeywordSchema>;
export type KeywordListResponse = z.infer<typeof KeywordListResponseSchema>;
export type Match = z.infer<typeof MatchSchema>;
export type MatchListResponse = z.infer<typeof MatchListResponseSchema>;
export type MonitorStatus = z.infer<typeof MonitorStatusSchema>;
export type CreateKeyword = z.infer<typeof CreateKeywordSchema>;

File: backend/internal/features/matches/schemas.go
package matches

type ListMatchesRequest struct {
	Page     int    `form:"page" binding:"omitempty,min=1"`
	PageSize int    `form:"page_size" binding:"omitempty,oneof=10 25 50"`
	Keyword  string `form:"keyword" binding:"omitempty,max=64"`
	Q        string `form:"q" binding:"omitempty,max=255"`
	Issuer   string `form:"issuer" binding:"omitempty,max=255"`
	DateFrom string `form:"date_from" binding:"omitempty"`
	DateTo   string `form:"date_to" binding:"omitempty"`
	NewOnly  string `form:"new_only" binding:"omitempty,oneof=true false"`
	Sort     string `form:"sort" binding:"omitempty,oneof=first_seen_desc last_seen_desc domain_asc"`
}

type ListMatchesResponse struct {
	Items []MatchItem `json:"items"`
	Total int         `json:"total"`
}

type MatchItem struct {
	ID           string  `json:"id"`
	KeywordID    string  `json:"keyword_id"`
	KeywordValue string  `json:"keyword_value"`
	CertSHA256   string  `json:"certificate_sha256"`
	MatchedField string  `json:"matched_field"`
	MatchedValue string  `json:"matched_value"`
	DomainName   *string `json:"domain_name"`
	IssuerCN     *string `json:"issuer_cn"`
	IssuerOrg    *string `json:"issuer_org"`
	SubjectCN    *string `json:"subject_cn"`
	SubjectOrg   *string `json:"subject_org"`
	NotBefore    *string `json:"not_before"`
	NotAfter     *string `json:"not_after"`
	FirstSeenAt  string  `json:"first_seen_at"`
	LastSeenAt   string  `json:"last_seen_at"`
	IsNew        bool    `json:"is_new"`
	CtLogIndex   int64   `json:"ct_log_index"`
}

type ErrorResponse struct {
	Error   string `json:"error"`
	Message string `json:"message,omitempty"`
}
